### PR TITLE
DR-2056: Create new API for faceted search

### PR DESCRIPTION
**Pull request by tcjiang9**
Opted to include all relevant data in this API so we can just use one API call for the catalogue

### CODE REVIEW

File: src/main/resources/api/data-repository-openapi.yaml [hunk 1]
```diff
@@ -2267,6 +2267,72 @@ paths:
 ##############################################################################
 ## SEARCH API STANDARD ENDPOINTS
 ##############################################################################
+  /api/repository/v1/search:
```
Comment (pshapiro4broad) [inline src/main/resources/api/data-repository-openapi.yaml]: I think this would be an API under `search` not at the root level. Maybe `/search/metadata` (to match the PUT and DELETE endpoints)?

File: src/main/resources/api/data-repository-openapi.yaml [hunk 2]
```diff
@@ -2267,6 +2267,72 @@ paths:
...
+    get:
+      tags:
+        - snapshots
+        - repository
+      description: >
+        Returns a list of all snapshots a user has access to and their search criteria
+      operationId: enumerateSnapshotSearch
+      parameters:
+        - name: offset
```
Comment (pshapiro4broad) [inline src/main/resources/api/data-repository-openapi.yaml]: I think for MVP we won't need any of these parameters, and I'd like to avoid adding any features until they're needed. So unless we need these immediately, I think it would be better to make this endpoint take no parameters.

File: src/main/resources/api/data-repository-openapi.yaml [hunk 3]
```diff
@@ -3465,6 +3531,34 @@ components:
...
         The total number of snapshots available that match the criteria and a page of summaries
+    EnumerateSnapshotSearchModel:
+      type: object
+      properties:
+        total:
+          type: integer
+          description: Total number of snapshots
+        filteredTotal:
+          type: integer
+          description: The number of snapshots included in this enumeration
```
Comment (pshapiro4broad) [inline src/main/resources/api/data-repository-openapi.yaml]: If we don't support pagination then we won't need these results.

File: src/main/resources/api/data-repository-openapi.yaml [hunk 4]
```diff
@@ -3465,6 +3531,34 @@ components:
...
+          type: array
+          items:
+            $ref: '#/components/schemas/SnapshotSearchModel'
+      description: >
+        The total number of snapshots available that match the criteria and a page of summaries
+    SnapshotSearchModel:
+      type: object
+      properties:
+        snapshot:
+          $ref: '#/components/schemas/SnapshotModel'
```
Comment (pshapiro4broad) [inline src/main/resources/api/data-repository-openapi.yaml]: I don't think this will return the snapshot itself. I'm not sure it will be necessary to return the snapshot ID either, as the ID will be a field in the metadata in the result, so it can be derived from the metadata.

**Review by fboulnois — CHANGES_REQUESTED** [review]
In addition to Phil's comments, I suggest adding dummy methods to the controller. When Phil merges in the db changes, you can also wire your methods up to the db methods.

File: src/main/resources/api/data-repository-openapi.yaml (hunk 4 above)
Comment (tcjiang9) [reply src/main/resources/api/data-repository-openapi.yaml]: what metadata model are you referring to?

Comment (pshapiro4broad) [reply src/main/resources/api/data-repository-openapi.yaml]: The one @fboulnois is working on. So in TIM this would be `dct:identifier` on the `TerraDCAT_ap:DataSnapshot` object. (It looks like TIM doesn't currently define `dct:identifier` for snapshots, but I think it wouldn't be controversial to add it.)

File: src/main/java/bio/terra/app/controller/SearchApiController.java [hunk 5]
```diff
@@ -80,6 +85,14 @@ private AuthenticatedUserRequest getAuthenticatedInfo() {
     return authenticatedUserRequestFactory.from(request);
   }
 
+  @Override
+  public ResponseEntity<Object> enumerateSnapshotSearch() {
+    List<UUID> authorizedResources =
+        iamService.listAuthorizedResources(getAuthenticatedInfo(), IamResourceType.DATASNAPSHOT);
+    Map<UUID, String> metadata = searchService.enumerateSnapshotSearch(authorizedResources);
+    return ResponseEntity.status(HttpStatus.OK).body(metadata);
```
Comment (pshapiro4broad) [inline src/main/java/bio/terra/app/controller/SearchApiController.java]: It's a minor thing but I think `.ok()` is a bit more convenient. Also I think we only want to return the `values()` since each metadata blob contains the uuid.
```suggestion
    return ResponseEntity.ok(metadata.values());
```

**Review by pshapiro4broad — APPROVED** [review]
Looks OK, but it should have a `MockMvc` test of the endpoint code. You can use the code in `SearchApiControllerTest` as a guide.

File: src/test/java/bio/terra/service/search/SearchServiceTest.java [hunk 6]
```diff
@@ -96,6 +96,16 @@ public void timColumnEncodingTest() {
     assertEquals(expectedSql, actualSql);
   }
 
+  @Test
+  public void testEnumerateSnapshotSearch() {
+    UUID testId = UUID.randomUUID();
+    List snapshotIds = List.of(testId);
```
Comment (pshapiro4broad) [inline src/test/java/bio/terra/service/search/SearchServiceTest.java]: Our code should avoid using generic types without arguments. You can use `var` to avoid typing the arguments, if you don't want to write `List<UUID>`. (In cases where the type is unknown you should use `List<?>`, not `List`.)

File: src/main/java/bio/terra/service/search/SearchService.java [hunk 7]
```diff
@@ -201,4 +212,11 @@ public SearchQueryResultModel querySnapshot(
     }
     return response;
   }
+
+  public SearchMetadataModel upsertSearchMetadata(UUID id, String body) {
```
Comment (pshapiro4broad) [inline src/main/java/bio/terra/service/search/SearchService.java]: Why did you move this out of the controller code? Moving it here doesn't seem to add any benefit.

File: src/main/java/bio/terra/service/search/SearchService.java [hunk 8]
```diff
@@ -201,4 +212,11 @@ public SearchQueryResultModel querySnapshot(
...
+    snapshotSearchMetadataDao.putMetadata(id, body);
+    SearchMetadataModel searchMetadataModel = new SearchMetadataModel();
+    searchMetadataModel.setMetadataSummary("Upserted search metadata for snapshot " + id);
+    return searchMetadataModel;
```
Comment (pshapiro4broad) [inline src/main/java/bio/terra/service/search/SearchService.java]: This should be done in the controller, since the service layer shouldn't need to know about the controller's models.

File: src/main/java/bio/terra/service/search/SearchService.java [hunk 9]
```diff
@@ -149,6 +155,11 @@ public SearchIndexModel indexSnapshot(Snapshot snapshot, SearchIndexRequest sear
     }
   }
 
+  public Map<UUID, String> enumerateSnapshotSearch(List<UUID> snapshotIDs) {
+    Map metadata = snapshotSearchMetadataDao.getMetadata(snapshotIDs);
```
Comment (pshapiro4broad) [inline src/main/java/bio/terra/service/search/SearchService.java]: I'm still unconvinced that it makes sense to have this API layer. If we need more machinery in the future, we can move it here, but for now having it here doesn't seem to fix a problem.

File: src/test/java/bio/terra/app/controller/SearchApiControllerTest.java [hunk 10]
```diff
@@ -38,33 +44,63 @@
...
   public void testUpsert() throws Exception {
+    var endpoint = "/api/repository/v1/search/{id}/metadata";
     var id = UUID.randomUUID();
-    var json = "{\"dct:identifier\": \"my snapshot\", \"dcat:byteSize\" : \"10000\"}";
-    mvc.perform(put(ENDPOINT, id).contentType(MediaType.APPLICATION_JSON).content(json))
-        .andExpect(status().isOk())
-        .andExpect(jsonPath("$.metadataSummary").value(containsString(id.toString())));
-    verifyAuthApi(id);
-    verify(snapshotMetadataDao).putMetadata(id, json);
+    var jsonString = "{\"dct:identifier\": \"my snapshot\", \"dcat:byteSize\" : \"10000\"}";
```
Comment (pshapiro4broad) [inline src/test/java/bio/terra/app/controller/SearchApiControllerTest.java]: What was the purpose of changing this variable name? In general, I prefer to not append types to names, e.g. prefer `snapshots` vs `snapshotList`, `json` vs `jsonString`.

File: src/test/java/bio/terra/app/controller/SearchApiControllerTest.java [hunk 11]
```diff
@@ -38,33 +44,63 @@
...
   @MockBean private SnapshotSearchMetadataDao snapshotMetadataDao;
 
-  private static final String ENDPOINT = "/api/repository/v1/search/{id}/metadata";
+  @MockBean private SearchService searchService;
+
+  @MockBean private SnapshotService snapshotService;
 
   @Test
   public void testUpsert() throws Exception {
+    var endpoint = "/api/repository/v1/search/{id}/metadata";
```
Comment (pshapiro4broad) [inline src/test/java/bio/terra/app/controller/SearchApiControllerTest.java]: Why did you change this from a constant to a local variable? Now this string is duplicated in two different places.

File: src/test/java/bio/terra/app/controller/SearchApiControllerTest.java [hunk 12]
```diff
@@ -38,33 +44,63 @@
...
             eq(IamResourceType.DATASNAPSHOT),
             eq(id.toString()),
             eq(IamAction.UPDATE_SNAPSHOT));
+    verify(snapshotMetadataDao).deleteMetadata(id);
+  }
+
+  @Test
+  public void testEnumerateSnapshotSearch() throws Exception {
+    var endpoint = "/api/repository/v1/search/metadata";
+    mvc.perform(get(endpoint)).andExpect(status().isOk());
```
Comment (pshapiro4broad) [inline src/test/java/bio/terra/app/controller/SearchApiControllerTest.java:92]: Testing that an endpoint returns `200` is not a very informative test. Instead I'd expect the test to verify that the SAM and DAO code is called, and that the DAO response is sent back via the endpoint API.

File: src/test/java/bio/terra/app/controller/SearchApiControllerTest.java [hunk 13]
```diff
@@ -38,33 +44,63 @@
...
+    mvc.perform(
+            post(endpoint, id.toString())
+                .contentType(MediaType.APPLICATION_JSON)
+                .content(TestUtils.mapToJson(searchIndexRequest)))
+        .andExpect(status().isOk());
+    verify(iamService)
+        .verifyAuthorization(any(), eq(IamResourceType.DATAREPO), any(), eq(IamAction.CONFIGURE));
+    verify(iamService)
+        .verifyAuthorization(
+            any(), eq(IamResourceType.DATASNAPSHOT), eq(id.toString()), eq(IamAction.READ_DATA));
```
Comment (pshapiro4broad) [inline src/test/java/bio/terra/app/controller/SearchApiControllerTest.java:114]: This test also needs to verify that the query was passed to the service layer.

File: src/test/java/bio/terra/service/search/SearchServiceTest.java [hunk 14]
```diff
@@ -123,6 +139,15 @@ public void indexSnapshotTest() throws Exception {
     assertEquals(indexName, searchIndexModel.getIndexSummary());
   }
 
+  @Test
+  public void upsertSearchMetadataTest() {
+    var id = UUID.randomUUID();
+    var body = "body";
+    SearchMetadataModel result = service.upsertSearchMetadata(id, body);
+    verify(snapshotSearchMetadataDao, times(1)).putMetadata(id, body);
```
Comment (pshapiro4broad) [inline src/test/java/bio/terra/service/search/SearchServiceTest.java]: `times(1)` is redundant as this is the default value

**Review by pshapiro4broad — CHANGES_REQUESTED** [review]

File: src/test/java/bio/terra/app/controller/SearchApiControllerTest.java (hunk 11 above)
Comment (tcjiang9) [reply src/test/java/bio/terra/app/controller/SearchApiControllerTest.java]: because we have more than one endpoint now in the other tests, I can move them all to constants.

File: src/main/java/bio/terra/service/search/SearchService.java (hunk 8 above)
Comment (tcjiang9) [reply src/main/java/bio/terra/service/search/SearchService.java:220]: Why even have a service layer? I moved it here so the API layer tests just cover auth and API response, and any logic is tested in the service layer

Comment (tcjiang9) [reply src/main/java/bio/terra/service/search/SearchService.java:220]: I was considering moving `snapshotSearchMetadataDao.deleteMetadata(id);` to the service layer also but it was only one line. The existence of the service layer implied that the controller interacts with the service which interacts with the DAO objects, so those can all be unit tested as their own layers

File: src/main/java/bio/terra/service/search/SearchService.java (hunk 9 above)
Comment (tcjiang9) [reply src/main/java/bio/terra/service/search/SearchService.java]: What is the purpose a service layer, why not do everything in the controller? I moved it here so the API layer tests just cover auth and API response, and any operations with the DAO is tested in the service layer

I was considering moving snapshotSearchMetadataDao.deleteMetadata(id); from the controller to the service layer also but it was only one line. The existence of the service layer implied that the controller interacts with the service which interacts with the DAO objects, so those can all be unit tested as their own layers

File: src/main/java/bio/terra/app/controller/SearchApiController.java [hunk 15]
```diff
@@ -47,7 +48,7 @@
   private final AuthenticatedUserRequestFactory authenticatedUserRequestFactory;
   private final SearchService searchService;
   private final SnapshotService snapshotService;
-  private final SnapshotSearchMetadataDao snapshotMetadataDao;
```
Comment (pshapiro4broad) [inline src/main/java/bio/terra/app/controller/SearchApiController.java:50]: I've been thinking about the repetition we have in our class names vs the package names. It does seem redundant to call the DAO `SnapshotSearchMetadataDao` when it's in package `search`. But we do have this convention so I thought I could drop it from the variables at least, as it doesn't add any useful information to the name. But that practice doesn't follow our conventions either...

**Review by pshapiro4broad — APPROVED** [review]
Looks good!
